### PR TITLE
remove get_val in serialization

### PR DESCRIPTION
**Pull request by PSeitz**
remove get_val in serialization and mark as unimplemented!()
replace get_val with iter in linear codec
remove MultivalueStartIndexRandomSeeker
replace MultivalueStartIndexIter with closure
Sample 100 values in linear codec
remove Mutexes

### CODE REVIEW

File: fastfield_codecs/src/linear.rs [hunk 1]
```diff
@@ -126,18 +126,21 @@ impl FastFieldCodec for LinearCodec {
...
         }
 
-        // let's sample at 0%, 5%, 10% .. 95%, 100%
-        let num_vals = column.num_vals() as f32 / 100.0;
-        let sample_positions = (0..20)
-            .map(|pos| (num_vals * pos as f32 * 5.0) as u64)
-            .collect::<Vec<_>>();
+        let limit_num_vals = column.num_vals().min(100_000);
+
+        let num_samples = 100;
```
Comment (fulmicoton) [inline fastfield_codecs/src/linear.rs:131]: I like the step_by solution. 

Can you change the column.num_vals() limit above to 100 too, to make proofreading even easier.

File: fastfield_codecs/src/linear.rs [hunk 2]
```diff
@@ -126,18 +126,21 @@ impl FastFieldCodec for LinearCodec {
...
+        let step_size = (limit_num_vals / num_samples).max(1); // 20 samples
+        let mut sample_positions_and_values: Vec<_> = Vec::new();
+        for (idx, val) in column.iter().step_by(step_size as usize).enumerate() {
```
Comment (fulmicoton) [inline fastfield_codecs/src/linear.rs]: nitpick
I think you could have swapoed `step_by` and `enumerate`, and avoided l.135.

File: src/fastfield/multivalued/writer.rs [hunk 3]
```diff
@@ -204,112 +203,68 @@ impl MultiValuedFastFieldWriter {
...
+) -> impl Iterator<Item = u64> + 'a {
+    let mut offset = 0;
+    doc_id_map
+        .iter_old_doc_ids()
+        .chain(std::iter::once(u32::MAX))
+        .map(move |old_doc| {
+            if old_doc == u32::MAX {
+                // sentinel value for last offset
+                return offset;
+            }
```
Comment (fulmicoton) [inline src/fastfield/multivalued/writer.rs:261]: wouldn't doing the chain below do the same thing?

```
 doc_id_map
        .iter_old_doc_ids()
        .map(move |old_doc| { ... })
        .chain(iter::once(offset))
```

**Review by fulmicoton — APPROVED** [review]
see comments inline

File: fastfield_codecs/src/linear.rs (hunk 1 above)
Comment (PSeitz) [reply fastfield_codecs/src/linear.rs:131]: I don't understand, 100 instead of 100_000?

File: src/fastfield/multivalued/writer.rs (hunk 3 above)
Comment (PSeitz) [reply src/fastfield/multivalued/writer.rs:261]: It would, but offset is already owned by the closure

File: fastfield_codecs/src/linear.rs (hunk 2 above)
Comment (PSeitz) [reply fastfield_codecs/src/linear.rs]: yes, that's better

**Pull request merged**